src/core_plugins/input_control_vis/public/components/editor/index_pattern_select.js

```diff
@@ -1,45 +1,125 @@
+import _ from 'lodash';
 import PropTypes from 'prop-types';
 import React, { Component } from 'react';
-import Select from 'react-select';
 
 import {
   EuiFormRow,
+  EuiComboBox,
 } from '@elastic/eui';
 
 export class IndexPatternSelect extends Component {
   constructor(props) {
     super(props);
 
-    this.loadOptions = this.loadOptions.bind(this);
+    this.state = {
+      isLoading: false,
+      options: [],
+      selectedIndexPattern: undefined,
+    };
   }
 
-  loadOptions(input, callback) {
-    this.props.getIndexPatterns(input).then((indexPatternSavedObjects) => {
+  componentWillMount() {
+    this._isMounted = true;
+  }
+
+  componentWillUnmount() {
+    this._isMounted = false;
+    this.debouncedFetch.cancel();
+  }
+
+  componentDidMount() {
+    this.fetchOptions();
+    this.fetchSelectedIndexPattern(this.props.indexPatternId);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.indexPatternId !== this.props.indexPatternId) {
+      this.fetchSelectedIndexPattern(nextProps.indexPatternId);
+    }
+  }
+
+  fetchSelectedIndexPattern = async (indexPatternId) => {
+    if (!indexPatternId) {
+      this.setState({
+        selectedIndexPattern: undefined
+      });
+      return;
+    }
+
+    const indexPattern = await this.props.getIndexPattern(indexPatternId);
+
+    if (!this._isMounted) {
+      return;
+    }
+
+    // TODO: handle case where index pattern no longer exists
+
+    this.setState({
+      selectedIndexPattern: {
+        value: indexPattern.id,
+        label: indexPattern.title,
+      }
+    });
+  }
+
+  debouncedFetch = _.debounce(async (searchValue) => {
+    const indexPatternSavedObjects = await this.props.getIndexPatterns(searchValue);
+
+    if (!this._isMounted) {
+      return;
+    }
+
+    // We need this check to handle the case where search results come back in a different
+    // order than they were sent out. Only load results for the most recent search.
+    if (searchValue === this.state.searchValue) {
       const options = indexPatternSavedObjects.map((indexPatternSavedObject) => {
         return {
           label: indexPatternSavedObject.attributes.title,
           value: indexPatternSavedObject.id
         };
       });
-      callback(null, { options: options });
-    });
+      this.setState({
+        isLoading: false,
+        options,
+      });
+    }
+  }, 300);
+
+  fetchOptions = (searchValue = '') => {
+    this.setState({
```

Comment:
Seems like `isLoading` might want to be dependent on *any* fetch which is happening, not just on `fetchOptions`. That'd require some coordination, but...